homeassistant/components/stream/worker.py

```diff
@@ -154,43 +152,42 @@ def check_flush_part(self, packet: av.Packet) -> None:
                 stream_id=self._stream_id,
                 init=self._memory_file.getvalue(),
             )
-            self._segment_last_write_pos = byte_position
+            self._memory_file_pos = self._memory_file.tell()
+            self._part_start_dts = self._segment_start_dts
             # Fetch the latest StreamOutputs, which may have changed since the
             # worker started.
             for stream_output in self._outputs_callback().values():
                 stream_output.put(self._segment)
         else:  # These are the ends of the part segments
-            self._segment.parts.append(
-                Part(
-                    duration=float(
-                        (packet.dts - self._part_start_dts) * packet.time_base
-                    ),
-                    has_keyframe=self._part_has_keyframe,
-                    data=self._memory_file.getbuffer()[
-                        self._segment_last_write_pos : byte_position
-                    ].tobytes(),
-                )
-            )
-            self._segment_last_write_pos = byte_position
-            self._part_start_dts = packet.dts
-            self._part_has_keyframe = False
+            self.flush_part(packet, last_part=False)
 
-    def flush(self, duration: Fraction, packet: av.Packet) -> None:
-        """Create a segment from the buffered packets and write to output."""
-        self._av_output.close()
+    def flush_part(self, packet: av.Packet, last_part: bool) -> None:
+        """Output a part from the most recent bytes in the memory_file."""
         assert self._segment
-        self._segment.duration = float(duration)
-        # Also flush the part segment (need to close the output above before this)
+        self._memory_file.seek(self._memory_file_pos)
         self._segment.parts.append(
             Part(
                 duration=float((packet.dts - self._part_start_dts) * packet.time_base),
                 has_keyframe=self._part_has_keyframe,
-                data=self._memory_file.getbuffer()[
-                    self._segment_last_write_pos :
-                ].tobytes(),
+                data=self._memory_file.read(),
             )
         )
+        if last_part:
+            self._memory_file_pos = 0
```

Comment:
It's a little cleaner, and we can leave a comment in just noting that we know about it being meaningless in the case of the last part.
Although the issue is really that memory_file_pos is used by the `Part`s, while memory_file is used by the `Segment`. So while they may seem like they should be executed at the same time because the names are similar and one is a reference into the other, there is a reason for the mismatch. It's really whether we want to "hide" this mismatch in a duplicated instruction or not.